### ISSUE TITLE
[libc++] Codecvts that decode UTF-8 can incorrectly return partial instead of error

### ISSUE THREAD

**Issue by dimztimz** [llvm/llvm-project#60177]
All codecvts that decode UTF-8 to other Unicode encodings (UTF-32, UTF-16 and UCS-2) have a bug in one case where they return `partial` instead of `error`. The case is when there is incomplete code sequence at the end of the input range but also in the visible trailing bytes there is some error. Here is example that reproduces the bug.

```cpp
#include <locale>
#include <cassert>

using namespace std;

int main()
{
	char in[]                 =  "bш\uAAAA\U0010AAAA";
	const char32_t expected[] = U"bш\uAAAA\U0010AAAA";
	char32_t out[ sizeof(expected)/sizeof(expected[0]) ];
	
	auto& loc = locale::classic();
	auto& cvt = use_facet<std::codecvt<char32_t, char, mbstate_t>>(loc);
	auto state = mbstate_t{};
	const char* in_next;
	char32_t* out_next;

	// \uAAAA is located at bytes in[3..5]
	auto orig = in[4];
	in[4] = 'z'; // we malform the second byte of the 3-byte sequence of \uAAAA

	// this call does not see the last byte of \uAAAA, it can only see its
	// first 2 bytes. If those 2 bytes were correct it should return partial,
	// but we malformed the second (and visible) byte. It should return error.
	auto res = cvt.in (state, in, in + 5, in_next, out, out + 5, out_next);
	assert(res == cvt.error); //error is correct, partial is returned

	in[4] = orig;

	// \U0010AAAA is located at bytes in[6..9]
	orig = in[7];
	in[7] = 'z'; // we malform the second byte of the 4-byte sequence of \U0010AAAA

	// this call does not see the last byte of \U0010AAAA, it can only see its
	// first 3 bytes. If those 3 bytes were correct it should return partial,
	// but we malformed the second (and visible) byte. It should return error.
	res = cvt.in (state, in, in + 9, in_next, out, out + 5, out_next);
	assert(res == cvt.error); //error is correct, partial is returned

	in[7] = orig;

	// can be also reproduced if we malform in[8]
}
```

**Comment by dimztimz** [comment]
> This looks like a duplicate of #14131. @dimztimz it looks like you're working on a fix, so could you check that?

It is a different bug. That bug is with `codecvt<wchar_t, char, mbstate_t>`, the codecvt that exists in C++98, this bug is with the newer codecvt's that were added in C++11.

**Comment by ldionne** [comment]
@dimztimz It looks like https://reviews.llvm.org/D143349 is ready to make progress -- it is blocked on you applying a few review comments but then we should be able to merge it and fix this!
